Add blog post on CHERI vs CHERIoT (not a thing) (#46)
CHERIoT-Platform/CHERIoT-Platform.github.io: _posts/2025-11-19-cheri-or-cheriot.markdown

---
layout: post
title:  "CHERI or CHERIoT?"
date:   2025-11-19
categories: cheri philosophy isa
author: "David Chisnall"
---

Recently, a few people have asked me 'should we do CHERI or CHERIoT?'

> I hadn't previously written an answer to this because the question doesn't make sense: it's like asking 'should we do x86 page tables or MMUs?'

Reviewer: Possibly flip these so they're parallel: `CHERI : MMU :: CHERIoT : X86 page tables`?

```suggestion
I hadn't previously written an answer to this because the question doesn't make sense: it's like asking 'should we do MMUs or x86 page tables?'
```

Since it's been asked several times, I think it's worth taking some time to explain *why* the question doesn't make sense.

CHERI is an abstract architecture
---------------------------------

> CHERI is a set of ISA extensions that define a capability model for accessing memory within an address space.

Reviewer: I know what you mean with "set of ISA extensions", but that still sounds kind of concrete -- "which ISA in particular?".  Perhaps something more obviously abstract, like...
```suggestion
CHERI is a conceptual extension to the von Neumann model of computation, defining a **capability model** for accessing resources (such as memory) within an address spaces.
```

> There are a *lot* of CHERI variants.

Reviewer: ```suggestion
There are a *lot* of ways to make CHERI more concrete.
```

The [original CHERI research was done on a 64-bit MIPS variant.](https://www.cl.cam.ac.uk/research/security/ctsrd/pdfs/201406-isca2014-cheri.pdf)
[Arm Morello extended ARMv8 with CHERI extensions](https://www.arm.com/architecture/cpu/morello).

> The University of Cambridge [CHERI ISA v9 Technical Report](https://www.cl.cam.ac.uk/techreports/UCAM-CL-TR-987.pdf) contains x86 and RISC-V CHERI adaptations.

Reviewer: Maybe find some phrasing that doesn't risk elevating x86 CHERI to the same level of understood and validated as RISC-V CHERI?

Software written for CHERI in C (or any language that is higher-level than C) doesn't care the bit patterns of capabilities, how bounds or permissions are encoded, and so on.
It cares only that you have some capability registers, that you can store capabilities in memory, and that the C language notion of a pointer can be represented by a CHERI capability for the target ISA.

I like to use memory-management units (MMUs) as an analogy for CHERI.
MIPS and SPARC had software-managed translation-lookaside buffers (TLBs).
x86 and Arm had architectural page tables that hardware walked to fill the TLB automatically on miss.
PowerPC had an MMU design that, in the words of Linus Torvalds, 'can be used to frighten small children'.
All of these implemented the same set of abstractions in terms of processes, shared memory, copy-on-write regions, and so on.
All of these are MMU designs, just as there are many possible CHERI implementations.

There are lots of things like this in ISAs.
Arm's Neon, PowerPC's AltiVec, and x86 SSE are all SIMD extensions, which provide similar programming models, but they all have quite different low-level details.

> It's quite possible to write code that targets all of them.

Reviewer: Especially with compiler assistance.  (It's also possible to hand-write or otherwise generate assembler for a specific instance, and people admittedly _do_, chasing every bit of performance.)

CHERIoT is a concrete platform
------------------------------

CHERIoT is a complete ISA specification, co-designed with a software stack.
In the same way that x86 page tables or MIPS' software-managed TLB are both concrete instances of the abstract idea of a memory-management unit, CHERIoT is a concrete instantiation of CHERI.

> 

Reviewer: ```suggestion
(As are Morello, Cambridge's original CHERI MIPS, and Cambridge's current ISAv9 CHERI-RISC-V.)

```

This means that there is no 'CHERI or CHERIoT' choice.
The real choices are between CHERI or not CHERI, or between CHERIoT and some other concrete instantiation of CHERI.
Having worked on CHERI since 2012, I am obviously biased on the first of these: given a choice between CHERI and not-CHERI, you should definitely pick CHERI!
The second choice (CHERIoT or some other CHERI variant) is more nuanced, as I'll discuss more in the rest of this post.
There are good reasons for CHERIoT but there are also domains where it is not the best choice.

CHERIoT is a *rich* CHERI platform
----------------------------------

A minimal CHERI specification provides pointer integrity, bounds enforcement, and some permissions.

> Pointer integrity means, among other things, that you can precisely differentiate between pointers and other things in memory and so can build temporal safety on top.

Reviewer 2: Pointer integrity means, among other things, that pointers can be reliably distinguished from non-pointer data in memory, allowing you to build temporal safety on top.

Trying to avoid "other things" mentioned twice.

> Pointer integrity means (among other things) that you can precisely differentiate between pointers and other kinds of non-pointer data in memory and so can build temporal safety on top.

Reviewer: ```suggestion
Pointer integrity means (among other things) that you can precisely differentiate between pointers and other kinds of non-pointer data in memory.
This is sufficient to detect attempts to exploit pointer/integer type confusion (as in traditional buffer overflows or ROP techniques) and is a central aspect of CHERI temporal safety technologies.

```

> Most of the CHERI temporal safety work has given protection against use-after-reallocation issues.

Reviewer 2: Just an opinion:
Most of the CHERI temporal safety work has provided us protection against use-after-reallocation.

> An pointer in a CHERI system may point to deallocated memory, but it will not point to *reallocated* memory.

Reviewer 2: A pointer.

> A pointer in a CHERI system may point to deallocated memory, but it will not point to *reallocated* memory.

Reviewer: ```suggestion
A pointer in a CHERI system with Cornucopia-style heap temporal safety may point to `free()`-ed memory, but it will not point to *reallocated* memory.
```

Something in the system will ensure that deallocated pointers are gone before the memory allocator reuses memory.
This means that any use-after-free bug will either point to the old object, or will trap.
It will never point to a new object and cause accidental aliasing.

> 
> CHERIoT builds on top of the core CHERI ideas with a couple of hardware features that guarantee deterministic trapping for use-after-free bugs.

Reviewer: ```suggestion
By contrast, CHERIoT takes advantage of its small memory sizes to build, on top of the core CHERI ideas, a couple of hardware features that guarantee deterministic trapping for use-after-free bugs.
That is, as soon as a `free()` returns, every pointer to the freed object -- even interior pointers and those held by other threads -- will trap if dereferenced.

```

It extends the sentry mechanism in CHERI systems [to provide stronger control-flow integrity properties](/isa/ibex/2024/06/26/sentries-cfi.html) than most other CHERI platforms, and also uses them to enforce interrupt control.

The entire system is designed to allow [fine-grained auditing of compartment rights](/rtos/firmware/auditing/2024/03/01/cheriot-audit.html), which lets us respect the principle of least privilege by having compartments as small as a single function *and statically reasoning about what they can do in the presence of a compromise*.
This set of building blocks lets us do things like [privilege separate the network stack and make callers respect the principle of intentional use, again with auditable guarantees that let you reason about which compartments can talk to which remote servers](/rtos/networking/auditing/2024/03/08/cheriot-network-stack.html).

This adds up to an easy-to-use programmer model and richer security guarantees than, we believe, any other system (CHERI or otherwise).

CHERIoT is designed for resource-constrained systems
----------------------------------------------------

> Some of the strengths of CHERIoTs come from the fact that we were willing to tailor the system to small devices.

Reviewer 2: CHERIoTs -> CHERIoT systems?

Prior CHERI systems achieve temporal memory safety by using the MMU to allow concurrent revocation.

> CHERIoT was designed to not need an MMU and so needed another mechanism for temporal safety, inspired by our previous CHERI+MTE work.

Reviewer 2: This took a double take. I know the CHERI/CHERIoT differences. This seemed a bit abrupt.  I'd almost be tempted to remove "Prior CHERI systems achieve temporal memory safety by using the MMU to allow concurrent revocation."

Author: If I remove that, it's not clear why it's a problem to not have an MMU for CHERIoT.

Reviewer 2: I know why it didn't read right. It was like saying "CHERIoT was designed not to need an MMU" but because of that it needed an alternative mechanism. When it's the alternative mechanism that means an MMU isn't needed.

Prior CHERI systems rely on the MMU to support concurrent revocation as their mechanism for temporal memory safety. CHERIoT does not include or need an MMU, it uses an alternative approach for temporal safety, inspired by our previous CHERI+MTE work.

Page tables are large data structures and they enforce a minimum granule size for memory protection and accounting.
A minimal process isolated with a typical MMU needs one page for the stack, one page of code, and one for read-write globals.
On top of that, it needs a complete page table, which is at least two pages.
This means you need a minimum of 20 KiB for an isolated component, of which 8 KiB is purely book-keeping overhead.
On a CHERIoT system, a lot of our compartments are smaller than the page tables that would be necessary to isolate them.
In addition, MMUs bring nondeterminism, which is not ideal in any system with even fairly soft realtime requirements.

The temporal memory system in CHERIoT is designed to scale across the range of microcontroller cores, including low-core-count multicore systems.
We check whether capabilities are valid when you load them.
Most pointers are used more than once, so this is more efficient than checking on every use (as an MTE system must do), but does not scale as well to large multicore systems or complex memory hierarchies.
You would not want to build a temporal-safety system like ours for a large server system, for example.

> CHERIoT also uses sentries for interrupt control.

Reviewer: ```suggestion
CHERIoT also extends CHERI's sealed function pointer ("sentry") mechanism for interrupt control.
```

What about standard RISC-V CHERI?
---------------------------------

The RISC-V standardisation process has an unenviable task of trying to create a standard base for RISC-V CHERI systems that will work for 2-stage microcontrollers, 25-stage superscalar out-of-order server chips, GPUs, AI accelerators, SmartNICs, and so on.
This is likely to end up being a (hopefully small) family of bases and a set of extensions.

If all goes well, CHERIoT v2 will be one of those bases plus some extensions.
RVB26 will be a RISC-V profile assembled from those bases and a set of extensions necessary to run a CHERI Linux or CheriBSD.

As with the rest of RISC-V, any useful chip implements a base architecture and a set of extensions.
Profiles exist to corral a set of extensions that software can depend on in environments where binary compatibility is important, such as those running off-the-shelf operating systems and programs.

So when should I use CHERIoT vs some other CHERI?
-------------------------------------------------

If you are looking for a microcontroller-class system, CHERIoT is probably the right answer.

> It is co-designed with a compartmentalisation model, includes a rich set of software abstractions, and provides temporal safety as a baseline feature.

Reviewer: ```suggestion
It is co-designed with a compartmentalisation model and RTOS implementation which includes a rich set of software abstractions and provides heap temporal safety as a baseline feature (and lightweight mechanism for defense against stack temporal safety violations).
```
